velox/common/base/SimdUtil-inl.h

```diff
@@ -1436,4 +1436,189 @@ inline bool memEqualUnsafe(const void* x, const void* y, int32_t size) {
+
+      if (memcmp(s + i + bitpos + 1, needle + 1, k - 2) == 0) {
+        return i + bitpos;
+      }
+
+      mask = detail::clearLeftmostSet(mask);
+    }
+  }
+
+  return std::string::npos;
+}
+
+template <size_t k, typename MEMCMP>
+size_t FOLLY_ALWAYS_INLINE smidStrstrMemcmp(
+    const char* s,
+    size_t n,
+    const char* needle,
+    MEMCMP memcmp_fun) {
+  assert(k > 0);
+  assert(n > 0);
+
+  auto first = CharVector::broadcast(needle[0]);
+  auto last = CharVector::broadcast(needle[k - 1]);
+  for (size_t i = 0; i < n; i += CharVector::size) {
+    auto block_first = CharVector::load_unaligned(s + i);
+    auto block_last = CharVector::load_unaligned(s + i + k - 1);
+
+    const auto eq_first = (first == block_first);
+    const auto eq_last = (last == block_last);
+
+    auto mask = toBitMask(eq_first && eq_last);
+
+    while (mask != 0) {
+      const auto bitpos = detail::getFirstBitSet(mask);
+
+      if (memcmp_fun(s + i + bitpos + 1, needle + 1)) {
+        return i + bitpos;
+      }
+
+      mask = detail::clearLeftmostSet(mask);
+    }
+  }
+
+  return std::string::npos;
+};
+} // namespace detail
+#endif
+
+/// A faster implementation for c_strstr(), about 2x faster than string_view`s
+/// find(), proved by TpchLikeBenchmark. Use xsmid-batch to compare first&&last
+/// char first, use fixed-memcmp to compare left chars. Inline in header file
+/// will be a little faster.
+size_t FOLLY_ALWAYS_INLINE
+simdStrstr(const char* s, size_t n, const char* needle, size_t k) {
+#ifdef SIMD_STRSTR
+  size_t result = std::string::npos;
+
+  if (n < k) {
+    return result;
+  }
+
+  switch (k) {
+    case 0:
+      return 0;
+
+    case 1: {
+      const char* res = strchr(s, needle[0]);
+
+      return (res != nullptr) ? res - s : std::string::npos;
+    }
+
+    case 2:
+      result = detail::smidStrstrMemcmp<2>(s, n, needle, alwaysTrue);
+      break;
+
+    case 3:
+      result = detail::smidStrstrMemcmp<3>(s, n, needle, memcmp1);
+      break;
+
+    case 4:
+      result = detail::smidStrstrMemcmp<4>(s, n, needle, memcmp2);
+      break;
+
+    case 5:
+      // Note: use memcmp4 rather memcmp3 for align, as the last character
+      // of needle is already proven to be equal
+      result = detail::smidStrstrMemcmp<5>(s, n, needle, memcmp4);
+      break;
+
+    case 6:
+      result = detail::smidStrstrMemcmp<6>(s, n, needle, memcmp4);
+      break;
+
+    case 7:
+      result = detail::smidStrstrMemcmp<7>(s, n, needle, memcmp5);
+      break;
+
+    case 8:
+      result = detail::smidStrstrMemcmp<8>(s, n, needle, memcmp6);
+      break;
+
+    case 9:
+      // Note: use memcmp8 rather memcmp7 for the same reason as above.
+      result = detail::smidStrstrMemcmp<9>(s, n, needle, memcmp8);
+      break;
+
+    case 10:
+      result = detail::smidStrstrMemcmp<10>(s, n, needle, memcmp8);
+      break;
+
+    case 11:
+      result = detail::smidStrstrMemcmp<11>(s, n, needle, memcmp9);
+      break;
+
+    case 12:
```

Comment:
You should extend it to 18, since 16 bytes fixed width comparison can be done on SSE in one instruction

Reply:
Still need to do this , If I removed FixedMemCmp for they affect performance a little. 

For memcpy , also affect a little, however they are cheap to add, and I not remove it.

Reply:
tpchQuery2                                                 10.35ms     96.60
tpchQuery9                                                 16.38ms     61.04
tpchQuery13                                                28.56ms     35.01
tpchQuery14                                                 9.28ms    107.70
tpchQuery16Part                                             8.29ms    120.58
tpchQuery16Supplier                                        25.18ms     39.71
tpchQuery20                                                 7.31ms    136.84